### DISCUSSION TITLE
Odin 2 zone on/off system

### DISCUSSION THREAD

**Discussion by JHAvB** [gekkekoe/esphome-ecodan-hp#384] [Odin - Q&A] [answered]
Over the past 9 days, I have been testing Odin and I can definitely see a lot of potential in the system. However, I still have a few questions regarding its operation within a two-zone system.

My system consists of a 2-zone on/off system, where both zones use the same supply temperature. However, the zones have very different characteristics:

Zone 1
•	Very high thermal mass / high heat capacity.
•	Relatively little glazing and therefore limited solar gains.
•	Heats up relatively slowly.
•	Emission system: fan coils with relatively high cooling capacity compared to the actual cooling load.
What I observe:
Odin activates IN1, while this zone remains structurally too cold and there is no cooling demand. The temperature remains consistently below the defined margin.

Zone 2
•	High thermal mass / reasonable heat capacity.
•	Relatively large glazing area and therefore significant solar gains.
•	Heats up considerably faster.
•	Emission system: cooling coil in the ventilation system (top cooling), where the available cooling capacity is smaller than the actual cooling demand.

What I observe:
Odin switches off IN6 during the night, while there is still a cooling demand and the room temperature remains above the defined margin.

My question:
Am I correct in understanding that Odin currently does not evaluate the individual zone inputs independently, but mainly controls based on the overall system demand? Or should Odin take into account the individual temperature and demand of each zone separately?

Additionally, I have a question regarding the correct heating system profile.
My emission systems (both the fan coils and the cooling coil) are designed for a water temperature range of 6°C supply temperature and 12°C return temperature at full cooling load.

I initially assumed that the UFH (underfloor heating) profile would be the closest match regarding temperature delta, but I am unsure whether this profile also influences the system dynamics/speed (for example, how quickly Odin responds or the expected thermal response).
Which profile would you recommend for a system with low-temperature cooling, fan coils, and a cooling coil?

**Comment by gekkekoe** [comment]
Hi!

Thanks for using Odin. Currently Odin uses zone1 room temp only to model the heat loss. The idea is that this is normally the largest area with UFH, the rest follows (z1 flow == z2 flow or z2 flow = mixed with z1). With this in mind, it does not make a lot of sense to model z2 also.

Can you tell me more about your system?
- buffer tank (probably needed by the fancoils)?
- mixing valve ?

Can your z2 generate demand on it's own? If so then I need to model z2. But this assumes there's a buffer tank since z2 probably can't handle the full capacity of the heatpump

Do you have some logging entries for me when Odin switches off in6? If needed we can adjust Odin for multi zones, but need to verify first since I don't want to add unneeded code.

Can you try Odin 1.0.5-01 from the release page, it has some improvements to reward night cooling. In the night it's more expensive in general, so it tries to avoid it, unless you are out of your comfort bandwidth

I assume UFH is the best profile.

**Comment by JHAvB** [comment]
Thanks for the very quick response.

It is clear now that ODIN currently only uses zone 1 for the heat loss model. However, I still wonder why IN1 is being activated while zone 1 is consistently well below the comfort range and there is no cooling demand from this zone. The actual cooling demand is coming from zone 2, so activating IN1 does not seem to contribute to the cooling demand.

Regarding my system:

The buffer tank is installed in series in the return pipe and is used as a system volume enlarger, not as an open buffer or hydraulic separator. I do not have additional circulation pumps or a mixing valve. To guarantee the minimum required flow rate of the heat pump, a bypass valve is installed in the system.

The cooling capacity of the cooling coil in zone 2 is highly dependent on the supply air temperature. A supply air temperature above approximately 18°C is already sufficient to release the minimum capacity of the heat pump.

In practice, only during the night, when the bypass around the heat recovery ventilation unit is open and the outdoor temperature drops below approximately 18°C, the heat pump starts to cycle because the available cooling demand becomes too low compared to the minimum output of the heat pump.

During daytime operation, zone 2 usually has sufficient cooling demand due to solar gains and the limited capacity of the cooling coil.

Where can I find the latest firmware version? Is this available on a different release page?

**Reply by gekkekoe** [reply]
I have sent the release info per email.

So If I understand it correctly
- z1 and z2 have the same flow temp? No mixing valve?
- z2 cannot generate demand on it's own?

In this case, only z1 probably needs to be modeled.

It would be helpful if I could get some logging from Asgard when it is starting to cycle. +1 hour before and couple of short cycles. Keep the browser to Asgard open (with logging tab selected) and there's a download button to collect the logs. Please send them to me via email. I can imagine that you need to set min flow range differently for z1 and z2. Probably it's hitting the min output as you suggested (and this might be due to the demand it computes from z1, but need logs to confirm)

Also in Odin you can download a debug json if you see in1 gets activated when it's still within comfort range (but also scroll a day ahead, it can do things to anticipate for tomorrow)
<img width="1302" height="582" alt="Screenshot 2026-07-13 at 14 57 17" src="https://github.com/user-attachments/assets/83d26ebd-512e-41fa-981a-b1ec290529c8" />
You can see it activating cooling at 00:90 to try to compensate for the out of band later today.

**Comment by JHAvB** [comment]
Thank you for sending the release!

I was away for a while, which is why I didn’t get back to you sooner. My apologies for that.

Since the release, it has been working a lot better. However, Z2 remains a bit of an odd one. I understand that you don’t want to run ODIN fully twice for Z2 as well. But would it be possible to force Z2 to be activated if it deviates significantly from the desired value?

ODIN is currently set for Z1 with a base temperature of 22.5 °C, with a minimum of -0.3 and a maximum of 0.3.

Would it be possible for IN6 to still be activated outside of ODIN if the Z2 error in cooling mode exceeds the minimum or maximum value twice?

For example: if the setpoint for Z2 is 23 °C, IN6 would be activated in cooling mode at, for example, 23.6 °C.

I’d be interested to hear your thoughts on this.

**Reply by gekkekoe** [reply]
This won't work if:
> z1 and z2 have the same flow temp? No mixing valve?
> z2 cannot generate demand on it's own?

From your previous logging:
> independent_zone_temps=0

This confirms that both zones have the same flow temp. In1 is requesting the demand for both zones. From my understanding z2 cannot generate demand on it's own? I am a bit confused since you mention in6, and then you could do things independently. But the logging states: independent_zone_temps=0. 

Is there a configuration error somewhere?

**Reply by JHAvB** [reply]
This is my situation:  
<img width="845" height="594" alt="image" src="https://github.com/user-attachments/assets/b578f341-2f89-46a9-8c21-dd05447ca6ac" />

When IN1 is activated by the Asgard hardware relay Z1, Ecodan valve 14 in the image is activated and opened. This valve controls the heating/cooling distribution for our living room.
When IN6 is activated by the Asgard hardware relay Z2, Ecodan valve 15 in the image is activated and opened.
Only Z2 can be activated, since valve 14 remains closed. However, in this two-zone system, there is only one supply temperature that is used for both zones, which is perfectly fine in my situation.  
Since the living room has more than enough cooling transfer capacity, Odin will only activate the heat pump occasionally for cooling. See also below.
<img width="945" height="197" alt="image" src="https://github.com/user-attachments/assets/069542e3-d856-4e9e-9520-f8f4cf8e4cf2" />

In my opinion, it would be useful if, when Z2 is, for example, 200% outside its comfort range as configured in ODIN, Asgard would only activate hardware relay Z2 and, in terms of control, operate the system at the minimum cooling capacity (2 kW in my case).
<img width="473" height="488" alt="image" src="https://github.com/user-attachments/assets/69ee2896-d49f-4437-baef-d5c61401fa32" />

**Reply by gekkekoe** [reply]
I am afraid it does not work like that. Just for verification: z2 can request demand on it's own? Without z1? What/who operates valve 16?

**Reply by JHAvB** [reply]
Yes, Z2 can operate completely independently without Z1. With Odin disabled, this is already working with Asgard.

16 is a bypass and not a valve. It serves as a safety measure to ensure the minimum flow rate. Both valve 14 and valve 15 have a switching time of approximately 3 minutes. Therefore, when opening, the bypass ensures the minimum flow rate.

**Reply by gekkekoe** [reply]
Ah thanks for the clarification. I need to look into ways to make this working for 2 zones. I don't have a fast solution and need to do some research on how to support this properly. Normally I always assume that z1 is needed.

In case of only z2 requesting demand, you could get into trouble with the min output (probably why you suggested 2kW, but that's only for your system, it has to work for all users).

Your z1 and z2 also have different characteristics, so we cannot share z1 learned data here and I really don't want to learn 2 parameter sets. z2 has cooling coils, is it the lossnay ?

**Reply by JHAvB** [reply]
Makes sense. That’s why I was also thinking about some kind of fallback in case Z2 deviates significantly from its setpoint. Normally, ODIN sends a power level for that hour, but perhaps it would be easier to fall back to the supply temperature calculated by AA.

I think some kind of override or fallback would be useful in any case. For example, if you want to heat or cool outside the scheduled period, that is currently not very easy to do.

Correct. Regarding the cooling in Z2, it is a cooling coil downstream of my Lossnay VL-350. Z2 also has some basic LT radiators for heating only.

**Reply by gekkekoe** [reply]
https://github.com/gekkekoe/esphome-ecodan-hp/discussions/297#discussioncomment-18235088

After a few days of researching, I made an implementation. The solution direction is a full 2nd training set. I need to do a lot of testing and hopefully I can release it for test this/next week.

<img width="624" height="123" alt="Screenshot 2026-09-01 at 16 23 25" src="https://github.com/user-attachments/assets/9628e10e-1938-4aa1-a6f7-550faeabc717" />

It's controlled via this switch. Off -> override to off (so old behavior)
Zone 1 first (default): Use z1 as main and z2 as residual
Zone 2 first: Use z2 as main and z1 as residual

@ethem I think you were also requesting this right?

**Reply by ethem** [reply]
Thanks for the ping @gekkekoe. Good to see the 2nd training set happening. For @JHAvB's setup that looks like the right answer.

For mine I'm not sure it applies, but I might be thinking about it wrong. What I was after in #363 was a second thermal mass rather than a second zone. My buffer sits in series in the primary circuit, upstream of the hydraulic separator. So it isn't a sink beside Z1, the heat Z1 uses comes out of it. Does the main/residual split still work in that case, or does the second set assume two zones drawing in parallel? And the buffer loses its heat to the plant room rather than to outside, so I'm not sure an AT-coupled parameter set fits it either.

One thing I'd sign onto regardless is JHAvB's point about an override for heating outside the plan. That's the same primitive I'd need for a charge block in a cheap price window, and probably a lot smaller than two-zone support.

**Reply by gekkekoe** [reply]
Each zone has a thermal mass, and is learned from z2 temperature. So don't know if this will work for you.

**Reply by JHAvB** [reply]
Thanks so much for the update! I’m looking forward to trying it out!

**Comment by gekkekoe** [comment]
You can try it with:
Asgard >= 2026.09.03-01
Odin >= 1.0.7-02
Documentation: https://github.com/gekkekoe/heatpump-optimizer

**Reply by JHAvB** [reply]
Hi Gekkekoe,

Thanks for releasing the update. Glad to see that Z2 is now also becoming part of Odin.
Over the past few hours, I noticed the following, but I'm not sure whether I have configured something incorrectly:
 
<img width="945" height="197" alt="image" src="https://github.com/user-attachments/assets/671b5eab-2d88-484c-9a43-5afa14fc5d6c" />

Normally, I don't see Expected Z2. I also don't see Z2 itself requesting demand from IN6. Yesterday, there were two moments when the cooling switched on, and in both cases, both IN1 and IN6 switched on.
 
<img width="945" height="193" alt="image" src="https://github.com/user-attachments/assets/6935a7ca-7928-4cab-bb25-b435b1bbefa6" />

As soon as I press Run optimization, I do see Expected Z2 for a few seconds, as well as a request for cooling, but this disappears again after a few seconds.
 
<img width="945" height="206" alt="image" src="https://github.com/user-attachments/assets/ab111acd-1444-4e14-88ab-ff254dfd3203" />

I'd be interested to hear whether this behavior is expected or if I have something configured incorrectly.

**Reply by gekkekoe** [reply]
Did some digging and I think I found it. Could you please post your dipswitches? (they are visible on http://esp-ip )
Currently, z2 solve via de background handler is triggered when zone temps are independent, but I think that this was not the case for you. 

SW2-7 is only valid active when SW3-6 is off. The background handler request probably single zone solve while the front-end request it with multi zone

**Reply by JHAvB** [reply]
Two different temperatures are indeed not possible in my system, because none of the zones has a mixing valve. According to the manual (unfortunately only available in Dutch), for my system, the setting for a “simple 2-zone valve ON/OFF control” is that DIP switch SW3-6 should be ON and SW2-7 should be OFF.

So, if I understand it correctly, this is indeed where things are going wrong.

<img width="954" height="218" alt="image" src="https://github.com/user-attachments/assets/45343459-931f-44a7-8c27-b734b8dd1d4a" />

<img width="747" height="469" alt="image" src="https://github.com/user-attachments/assets/818bfe79-4a1c-4f5c-88f7-1f47c277d787" />

**Reply by gekkekoe** [reply]
Ah ok then auto detect based on dipswitch is not possible, I will enable it for 2 zones in the background. You can use the new option in Odin to control how the z2 is solved.

Update to Asgard https://github.com/gekkekoe/esphome-ecodan-hp/releases/tag/2026-09-04.01 and it should work (predicted z2)

**Reply by JHAvB** [reply]
Thanks for the new release.

In the system logs, I can see that both Z1 and Z2 are requesting cooling from Odin.

However, somehow only IN1 becomes active, while IN6 does not.

<img width="945" height="193" alt="image" src="https://github.com/user-attachments/assets/53884143-51c3-486a-a20f-ef86c60f301f" />


<img width="945" height="196" alt="image" src="https://github.com/user-attachments/assets/623e5e71-50d4-4b5f-b7c4-38abb89984af" />

[logs_2026-09-04T11_32_47.txt](https://github.com/user-attachments/files/31833437/logs_2026-09-04T11_32_47.txt)

**Reply by gekkekoe** [reply]
can you also upload debug json (email, since it contains sensitive info) ? You should ensure via setpoint z2 that there's always demand. Odin will override it. Just like z1. Or Post your z1/z2 config screens here also.

**Reply by JHAvB** [reply]
<img width="658" height="623" alt="image" src="https://github.com/user-attachments/assets/47490d24-6289-45e8-9c02-21d1f45bef56" />

**Reply by gekkekoe** [reply]
Thanks I have found it (I hope, since I don't have an actual 2 zone system, and need to emulate). Please try https://github.com/gekkekoe/esphome-ecodan-hp/releases/tag/2026-09-04.02

**Reply by JHAvB** [reply]
Yes, IN6 is currently active!

According to the log, only Z1 should currently have no demand. However, it now appears that Z1 is still using the smallest delta (2°C). IN1 is also still active, even though it should be switched off according to the log.

[logs_2026-09-04T12_40_31.txt](https://github.com/user-attachments/files/31835624/logs_2026-09-04T12_40_31.txt)

<img width="1155" height="245" alt="image" src="https://github.com/user-attachments/assets/f5504d6f-35c3-4e21-bb7a-58b012d6a2c9" />

**Reply by gekkekoe** [reply, marked as answer]
try https://github.com/gekkekoe/esphome-ecodan-hp/releases/tag/2026-09-04.03, missed one condition

**Reply by JHAvB** [reply]
Over the past few days, I’ve been testing version 2026-09-04.03 and it seems to be working perfectly! It’s great that 2 zone capability is now possible in combination with Odin.

I was wondering: is there a possibility that Z2 will get its own 24h schedule profile in the future? That way, moving into winter, I could maintain a different temperature for the bedroom compared to the living room.

**Reply by gekkekoe** [reply]
great that it works properly. Per zone bandwidth is possible, but probably very low prio compare to other tasks
